processor/processor.go

```diff
@@ -2797,16 +2806,53 @@ func (proc *Handle) userTransformAndFilter(
 	}
 	// Send to custom transformer only if the destination has a transformer enabled
 	if transformationEnabled {
-		userTransformationStat := proc.newUserTransformationStat(sourceID, workspaceID, destination)
-		userTransformationStat.numEvents.Count(len(eventList))
-		proc.logger.Debug("Custom Transform input size", len(eventList))
+		noOfEvents := len(eventList)
+		utMirroringEnabled, utMirroringSanityChecks := proc.isUserTransformMirroringEnabled()
+		userTransformationStat := proc.newUserTransformationStat(sourceID, workspaceID, destination, false)
+		var userTransformationMirroringStat *DestStatT
+		userTransformationStat.numEvents.Count(noOfEvents)
+		if utMirroringEnabled {
+			userTransformationMirroringStat = proc.newUserTransformationStat(sourceID, workspaceID, destination, true)
+			userTransformationMirroringStat.numEvents.Count(noOfEvents)
+		}
+		proc.logger.Debug("Custom Transform input size", noOfEvents)
 
 		trace.WithRegion(ctx, "UserTransform", func() {
 			startedAt := time.Now()
+
+			if utMirroringEnabled {
+				go func() {
+					response := proc.transformerClients.UserMirror().Transform(ctx, eventList)
+					d := time.Since(startedAt)
+					userTransformationMirroringStat.transformTime.SendTiming(d)
+					userTransformationMirroringStat.numOutputSuccessEvents.Count(len(response.Events))
+					filtered := lo.GroupBy(response.FailedEvents,
+						func(event types.TransformerResponse) bool {
+							return event.StatusCode == reportingtypes.FilterEventCode
+						},
+					)
+					userTransformationMirroringStat.numOutputFailedEvents.Count(len(filtered[false]))
+					userTransformationMirroringStat.numOutputFilteredEvents.Count(len(filtered[true]))
+					if utMirroringSanityChecks != nil {
+						utMirroringSanityChecks <- response
+						close(utMirroringSanityChecks)
+					}
+				}()
+			}
+
 			response = proc.transformerClients.User().Transform(ctx, eventList)
 			d := time.Since(startedAt)
 			userTransformationStat.transformTime.SendTiming(d)
 
+			if utMirroringEnabled && utMirroringSanityChecks != nil {
+				go func() {
+					mirroredResponse := <-utMirroringSanityChecks
+					if diff, equal := response.Equal(&mirroredResponse); !equal {
+						proc.logger.Errorn("UserTransform sanity check failed", logger.NewStringField("diff", diff))
```

Comment:
This `diff` might leak sensitive information in the logs :thinking: 

@ktgowtham @lvrach we could upload the diffs in an S3 bucket. if overkilling (given that we would enable only for very short periods of time) we can leave as it is.

Reply:
Ideally, we should push them into a bucket that has:
1. Limited access
2. Access log enabled (as far as I am aware, it is disabled in most of the buckets due to high traffic, and thus cost)
3. Short retention (24 hours or less)

In addition, if it is not there already, we should add a flag that disables storing to S3 by default.

Reply:
https://rudderlabs.slack.com/archives/C02AASYS6R1/p1744706930275819